### PR TITLE
Support transactions in MessageEndpoint

### LINKED ISSUE #17: listener does not support jta
with the following code, the message is normally consumed and it should not:
```
    @Override
    @Transactional(Transactional.TxType.REQUIRES_NEW)
    public void onMessage(Message message) {
        try {
            log.info("Received blue message: " + message.getBody(String.class));
            QuarkusTransaction.setRollbackOnly();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
 ```

whereas, this code will trigger a redeliver, which is expected:
```
    @Override
    public void onMessage(Message message) {
...
        throw new RuntimeException("oops");
    } 
```

### PR DESCRIPTION
**Pull request by gastaldi**
- Fixes #17 

/cc @tadamski @vsevel

### CODE REVIEW

**Comment by vsevel** [comment]
hello. I tested this morning with your branch `transactional`. I don't see a change in behavior. I am testing with:

```
    @Override
    @Transactional(Transactional.TxType.REQUIRES_NEW)
    public void onMessage(Message message) {
        try {
            log.info("Received message: " + message.getBody(String.class) + "; tx is active = "
                    + QuarkusTransaction.isActive());
            if (message.getJMSRedelivered()) {
                log.info("accepting message on the second try");
            } else {
                log.info("failing the transaction");
                QuarkusTransaction.setRollbackOnly();
                // throw new RuntimeException("oops");
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
```

when I fail the tx with `QuarkusTransaction.setRollbackOnly();` messages get consumed normally.
when I fail the text with `throw new RuntimeException("oops");` messages stay pending then get redelivered.

this probably means that the delivery is not associated with the jta tx.

**Comment by gastaldi** [comment]
@vsevel does it work if you remove the `Transactional.TxType.REQUIRES_NEW`?

**Comment by vsevel** [comment]
not responding to your last question.
I see issues with the producer use case as well.
I added some persistence when sending the message:
```
    @Path("/sendcreate")
    @GET
    @Transactional
    public String sendcreate() {
        try (JMSContext context = factory.createContext()) {
            Queue myQueue = context.createQueue("jms.queue.X");
            JMSProducer producer = context.createProducer();
            String giftDescription = "hello" + new Date();
            santaClausService.createGift(giftDescription);
            producer.send(myQueue, giftDescription);
            return "sent and created " + giftDescription;
        }
    }
```

I can see the message being received and the row being persisted.
I was expecting an error saying that you can't enlist 2 non-xa resource managers.

> does it work if you remove the Transactional.TxType.REQUIRES_NEW?

```
    @Override
    // @Transactional(Transactional.TxType.REQUIRES_NEW)
    public void onMessage(Message message) {
        try {
            boolean active = QuarkusTransaction.isActive();
            log.info("Received message: " + message.getBody(String.class) + "; tx is active = " + active);
            // rollback(message);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
```

yes. is that what you wanted me to test?

**Comment by vsevel** [comment]
and this works as well:

```
    @Override
    // @Transactional(Transactional.TxType.REQUIRES_NEW)
    public void onMessage(Message message) {
        try {
            boolean active = QuarkusTransaction.isActive();
            log.info("Received message: " + message.getBody(String.class) + "; tx is active = " + active);
            rollback(message);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private void rollback(Message message) throws JMSException {
        if (message.getJMSRedelivered()) {
            log.info("accepting message on the second try");
        } else {
            log.info("failing the transaction");
            // QuarkusTransaction.setRollbackOnly();
            throw new RuntimeException("oops");
        }
    }
```

message gets rolled back, and reprocessed.

**Comment by gastaldi** [comment]
> yes. is that what you wanted me to test?

I mean to keep the `@Transactional` attribute but not `REQUIRES_NEW` (use `REQUIRED` which is the default)

**Comment by gastaldi** [comment]
> hello. I tested this morning with your branch `transactional`. I don't see a change in behavior. I am testing with:
> 
> ```
>     @Override
>     @Transactional(Transactional.TxType.REQUIRES_NEW)
>     public void onMessage(Message message) {
>         try {
>             log.info("Received message: " + message.getBody(String.class) + "; tx is active = "
>                     + QuarkusTransaction.isActive());
>             if (message.getJMSRedelivered()) {
>                 log.info("accepting message on the second try");
>             } else {
>                 log.info("failing the transaction");
>                 QuarkusTransaction.setRollbackOnly();
>                 // throw new RuntimeException("oops");
>             }
>         } catch (Exception e) {
>             throw new RuntimeException(e);
>         }
>     }
> ```
> 
> when I fail the tx with `QuarkusTransaction.setRollbackOnly();` messages get consumed normally. when I fail the text with `throw new RuntimeException("oops");` messages stay pending then get redelivered.
> 
> this probably means that the delivery is not associated with the jta tx.

You can't control the message delivery transaction using JTA if you are using `REQUIRES_NEW`. Use `REQUIRED` (preferred), `SUPPORTS` or `MANDATORY`

**Comment by gastaldi** [comment]
Also, because the delivery is managed by JTA, Artemis will always return a XASession, so it doesn't matter if you use `JMSContext` or `XAJMSContext`, as the behavior will be the same

**Comment by vsevel** [comment]
in the ejb spec only `REQUIRED` and `NOT_SUPPORTED` were allowed from what I remember.
in my last tests I used `REQUIRED`, but that does not change the outcome.unless I missed something, it does not seem there is enlistment.
do you see something else?

**Comment by gastaldi** [comment]
@vsevel yes, it works for me, as you can see in the `TransactionTest`

**Comment by vsevel** [comment]
I used `REQUIRED` on the consumer test, and that is OK. let me try again on the producer side.

File: runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/TransactionAwareMessageEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void beforeDelivery(Method method) throws ResourceException {
+        if (transacted) {
+            Arc.container().requestContext().activate();
+            QuarkusTransaction.begin();
+            try {
+                Transaction transaction = TransactionManager.transactionManager().getTransaction();
+                // Enlisting the resource so the message delivery is part of the transaction
+                // See https://jakarta.ee/specifications/connectors/2.1/jakarta-connectors-spec-2.1#transacted-delivery-using-container-managed-transaction
+                if (!transaction.enlistResource(xaResource)) {
```
Comment (gastaldi) [inline runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/TransactionAwareMessageEndpoint.java:40]: @vsevel this is where the enlistment happens

**Comment by vsevel** [comment]
can't git to work on the producer side (
``` 
    @Path("/sendcreate")
    @GET
    @Transactional
    public String sendcreate() {
        String giftDescription = "hello" + new Date();
        try (JMSContext context = factory.createContext()) {
            Queue myQueue = context.createQueue("jms.queue.com.lodh.jee.ArteTest.Bank.jms.bankref1.QueueB");
            JMSProducer producer = context.createProducer();
            producer.send(myQueue, giftDescription);
        }
        santaClausService.createGift(giftDescription);
        return "sent and created " + giftDescription;
    }
```

message gets sent anyway. didn't check on the insert.
and no complaining on having 2 non xa resource managers,
just like I said in https://github.com/quarkiverse/quarkus-ironjacamar/pull/25#issuecomment-1704793485

**Comment by vsevel** [comment]
this should fail and it does not:
``` 
    @Path("/sendcreate")
    @GET
    @Transactional(Transactional.TxType.REQUIRES_NEW)
    public String sendcreate() {
        String giftDescription = "hello" + new Date();
        try (JMSContext context = factory.createContext()) {
            Queue myQueue = context.createQueue("jms.queue.com.lodh.jee.ArteTest.Bank.jms.bankref1.QueueB");
            JMSProducer producer = context.createProducer();
            producer.send(myQueue, giftDescription);
        }
        // QuarkusTransaction.setRollbackOnly();
        santaClausService.createGift(giftDescription);
        return "sent and created " + giftDescription;
    }
```

we should have an error on registering 2 non xa resource managers

**Comment by vsevel** [comment]
same with 
```
    @Path("/sendcreate")
    @GET
    @Transactional
    public String sendcreate() {
        String giftDescription = "hello" + new Date();
        try (JMSContext context = factory.createContext()) {
            Queue myQueue = context.createQueue("jms.queue.com.lodh.jee.ArteTest.Bank.jms.bankref1.QueueB");
            JMSProducer producer = context.createProducer();
            producer.send(myQueue, giftDescription);
        }
        santaClausService.createGift(giftDescription);
        // QuarkusTransaction.setRollbackOnly();
        return "sent and created " + giftDescription;
    }
```

it passes and it should not

**Comment by vsevel** [comment]
but there is an enlistement, for instance this does not push any message (which is correct):

```
   @Path("/sendcreate")
    @GET
    @Transactional
    public String sendcreate() {
        String giftDescription = "hello" + new Date();
        try (JMSContext context = factory.createContext()) {
            Queue myQueue = context.createQueue("jms.queue.com.lodh.jee.ArteTest.Bank.jms.bankref1.QueueB");
            JMSProducer producer = context.createProducer();
            producer.send(myQueue, giftDescription);
        }
        santaClausService.createGift(giftDescription);
        QuarkusTransaction.setRollbackOnly();
        return "sent and created " + giftDescription;
    }
```

I need to check what is hapenning on the db side,
all for tonight. cheers.

**Comment by gastaldi** [comment]
@vsevel I improved the Artemis IT by adding a PostgreSQL DB and Panache. Tests look okay to me

**Comment by vsevel** [comment]
hello @gastaldi 
I did some more tests.
I am assuming non xa behavior, and it behaves as if it was supporting it. 
specifically those tests should fail (and do not): 
- sending and inserting in the same tx (required/required)
- receiving and inserting in the same tx (required/required)

there should be an error saying that you can't enlist 2 non-xa resource managers in the same tx
if there is no error, it is either:
- we are not running in the same tx (I don't think that is the case)
- both resource managers are xa
- one resource manager is xa, and nayarana enables LRCO under the cover

it is probably one of the last 2 situations. if so we should have the tx logs somewhere (e.g. `data\tx-object-store`) and a recovery manager?? and that would make the application stateful. we need to investigate that.
the default behavior for quarkus should assume the app to be stateless (12 factors, microservice, ...)
